actions/publish-image/action.yaml

```diff
@@ -0,0 +1,162 @@
+      <public-registry>/<public-repo>/<image>:<tag>.
+    type: string
+
+  public-username:
+    description: |
+      The username used to authenticate against the Public registry.
+    type: string
+
+  public-password:
+    description: |
+      The password used to authenticate against the Public registry.
+    type: string
+
+  push-to-prime:
+    description: |
+      Indicates whether the image should be pushed to the Prime container
+      registry.
+    default: true
+    type: boolean
+
+  prime-registry:
+    description: The container registry used for Prime images.
+    type: string
+
+  prime-repo:
+    description: |
+      The repository component in a fully qualified image. For reference:
+      <prime-registry>/<prime-repo>/<image>:<tag>.
+    type: string
+
+  prime-username:
+    description: |
+      The username used to authenticate against the Prime registry.
+    type: string
+
+  prime-password:
+    description: |
+      The password used to authenticate against the Prime registry.
+    type: string
+
+  make-target:
+    description: |
+      The make target used to build and push the container image.
+    default: push-image
+    type: string
+
+runs:
+  using: composite
+
+  steps:
+  # Login to all registries before starting the pushing process.
+  # Short-circuit if either fails. This should decrease the likelihood
+  # of only one registry getting updated while the other fails.
+  - name: Login to registry [Public]
+    if: ${{ inputs.push-to-public }}
+    uses: docker/login-action@v3
+    with:
+      registry: ${{ inputs.public-registry }}
+      username: ${{ inputs.public-username }}
+      password: ${{ inputs.public-password }}
+
+  - name: Login to registry [Prime]
+    if: ${{ inputs.push-to-prime }}
+    uses: docker/login-action@v3
+    with:
+      registry: ${{ inputs.prime-registry }}
+      username: ${{ inputs.prime-username }}
+      password: ${{ inputs.prime-password }}
+
+  - name: Setup QEMU
+    uses: docker/setup-qemu-action@v3
+  - name: Setup Docker Buildx
+    uses: docker/setup-buildx-action@v3
+  - name: Install Cosign
+    uses: sigstore/cosign-installer@v3.5.0
+
+  - name: Build and push image [Public]
+    shell: bash
+    if: ${{ inputs.push-to-public }}
+    run: |
+      make ${{ inputs.make-target }}
+    env:
+      TAG: ${{ inputs.tag }}
+      TARGET_PLATFORMS: ${{ inputs.platforms }}
+      REPO: ${{ inputs.public-registry }}/${{ inputs.public-repo }}
+
+  - name: Build and push image [Prime]
+    shell: bash
+    if: ${{ inputs.push-to-prime }}
+    run: |
+      IID_FILE=$(mktemp)
+      export IID_FILE_FLAG="--iidfile ${IID_FILE}"
+      
+      make ${{ inputs.make-target }}
+      cosign sign --oidc-provider=github-actions --yes "${REPO}/${{ inputs.image }}@$(head -n 1 ${IID_FILE})"
+    env:
+      TAG: ${{ inputs.tag }}
+      TARGET_PLATFORMS: ${{ inputs.platforms }}
+      REPO: ${{ inputs.prime-registry }}/${{ inputs.prime-repo }}
```

Comment:
Is it possible to build the image once and push it in a separate step so we can avoid building the same image twice?
The Makefile below wouldn't work, but it roughly shows the idea
```make
.PHONY: buildx-machine
buildx-machine: ## create rancher dockerbuildx machine targeting platform defined by DEFAULT_PLATFORMS.
	@docker buildx ls | grep $(MACHINE) || docker buildx create --name=$(MACHINE) --platform=$(DEFAULT_PLATFORMS)

.PHONY: build-image 
build-image: buildx-machine
	docker buildx build -f Dockerfile \
		--builder=$(MACHINE) $(IID_FILE_FLAG) $(BUILDX_ARGS) \
		--platform=$(DEFAULT_PLATFORMS) -t $(REPO):$(TAG) \
		--output type=image
		
.PHONY: push-image
push-image: buildx-machine ## build the container image targeting all platforms defined by TARGET_PLATFORMS and push to a registry.
        for platform in ${TARGET_PLATFORMS}; do \
	        docker push $(REPO):$(TAG)-${platform}
	done
	@echo "Pushed $(REPO):$(TAG)"
```

Reply:
By default the second `make push-image` will use all layers from cache automatically, unless the `Dockerfile` was really badly structured. On the `security-scan` example:

First execution takes `~2m`:
```
2024-07-25T17:01:15.7574061Z [36;1mmake push-image[0m
...
2024-07-25T17:03:14.1367447Z #72 DONE 18.3s
2024-07-25T17:03:14.1733212Z Pushed quay.io/***/security-scan:v0.0.32
```

Second takes `~6s`:
```
...
2024-07-25T17:03:14.1770165Z [36;1mmake push-image[0m
...
2024-07-25T17:03:21.1387600Z #69 DONE 5.4s
2024-07-25T17:03:21.1873599Z Pushed ***/***/security-scan:v0.0.32
```
From [full log](https://productionresultssa7.blob.core.windows.net/actions-results/355283d7-fd81-4dea-98e6-796f9979d3ab/workflow-job-run-aa9ced9a-b916-5d4b-6d68-014377571ac1/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-07-26T20%3A56%3A16Z&sig=7WMZ0bbMBHZseglxsSvC3cvzBAizaUwusq8p5f%2BaJdY%3D&ske=2024-07-27T07%3A49%3A54Z&skoid=ca7593d4-ee42-46cd-af88-8b886a2f84eb&sks=b&skt=2024-07-26T19%3A49%3A54Z&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skv=2024-05-04&sp=r&spr=https&sr=b&st=2024-07-26T20%3A46%3A11Z&sv=2024-05-04).

The one thing to consider though is that if a specific deployment needs to push several [tags](https://github.com/rancher/security-scan/blob/f99af55ca4152bb486e371d7e2035d4b18f8874b/.github/workflows/release.yml#L17-L28), we may want to force them to run sequentially with `max-parallel: 1`.

This is mostly for the users of the GHA, so does not really affect this PR, but my theory is that we would avoid pushing the layers in parallel. So for example, if the multi-arch version is uploaded first, the next two executions should be faster to run (as they would be cached) and to push (as the layers are already at the target registry). So although both cases would take `~2m` to finish, running in parallel would cost about `~6m` of compute, versus a tiny bit more than `~2m`.